### PR TITLE
TCP handshake and port reuse fix

### PR DESCRIPTION
**Pull request by DC3-DCCI**
This is a replacement for an earlier pull request fix #91 that better detects TCP connection closure and better handles unexpected sequence numbers.

### CODE REVIEW

File: lib/dshell.py [hunk 1]
```diff
@@ -697,14 +699,27 @@ def TCP(self, addr, tcp, ts, **kwargs):
...
         try:
             # close connection
-            if tcp.flags & (dpkt.tcp.TH_FIN | dpkt.tcp.TH_RST):
-                conn = self.find(addr)
+            conn = self.find(addr)
+            if tcp.flags & (dpkt.tcp.TH_FIN | dpkt.tcp.TH_RST) and conn:
+                conn.closeIP(addr[0]) #track if FIN has been seen in connection
+
+            if conn:
+                if conn.connectionClosed(): #if FIN has been seen from both sides, trigger old logic
```
Comment (dev195) [inline lib/dshell.py]: Do we still need the self.clientclosed and self.serverclosed attached to the decoder object? It looks like all they're doing is standing in for the conn.connectionClosed() function at this point.

File: lib/dshell.py [hunk 2]
```diff
@@ -697,14 +699,27 @@ def TCP(self, addr, tcp, ts, **kwargs):
...
+                    self.clientclosed = True
+                    self.serverclosed = True
```
Comment (dev195) [inline lib/dshell.py]: Do we still need the self.clientclosed and self.serverclosed attached to the decoder object? It looks like all they're doing is standing in for the conn.connectionClosed() function at this point.

**Comment by DC3-DCCI** [comment]
The pull request has been updated based on your feedback.  The unnecessary variables were removed from the decoder object and are now only in the connection object.

**Comment by dev195** [comment]
Okay, looks good. Thanks for the fix!

**Pull request merged**